tests/wpunit/CustomPostTypeTest.php

```diff
@@ -509,4 +511,629 @@ public function testQueryCustomPostTypeWithSameValueForGraphqlSingleNameAndGraph
+			'graphql_exclude_connections' => [ 'comments', 'revisions' ],
+		]);
+
+		$query = '
+		{
+			excludedConnections {
+				nodes {
+					__typename
+					databaseId
+					comments {
+						nodes {
+							__typename
+						}
+					}
+					revisions {
+						nodes {
+							__typename
+						}
+					}
+				}
+			}
+		}
+		';
+
+		$actual = $this->graphql( [ 'query' => $query ] );
+
+		$this->assertArrayHasKey( 'errors', $actual );
+		$this->assertStringStartsWith( 'Cannot query field "comments"', $actual['errors'][0]['message'] );
+		$this->assertStringStartsWith( 'Cannot query field "revisions"', $actual['errors'][1]['message'] );
+
+		unregister_post_type( 'missing_connections' );
+	}
+
+	public function testRegisterCustomPostTypeWithGraphQLKindNoResolver() {
+
+		$this->tester->expectThrowable( \Exception::class, function () {
+			register_post_type( 'with_interface_kind', [
+				'public'              => true,
+				'show_in_graphql'     => true,
+				'graphql_single_name' => 'WithInterfaceKind',
+				'graphql_plural_name' => 'WithInterfaceKinds',
+				'graphql_kind'        => 'interface',
+			]);
+
+		} );
+
+		$this->tester->expectThrowable( \Exception::class, function () {
+			register_post_type( 'with_union_kind_one', [
+				'public'              => true,
+				'show_in_graphql'     => true,
+				'graphql_single_name' => 'WithUnionKind',
+				'graphql_plural_name' => 'WithUnionKinds',
+				'graphql_kind'        => 'union',
+			]);
+
+		} );
+
+		$this->tester->expectThrowable( \Exception::class, function () {
+			register_post_type( 'with_union_kind_two', [
+				'public'               => true,
+				'show_in_graphql'      => true,
+				'graphql_single_name'  => 'WithUnionKind',
+				'graphql_plural_name'  => 'WithUnionKinds',
+				'graphql_kind'         => 'union',
+				'graphql_resolve_type' => $this->resolve_type(),
+			]);
+
+		} );
+
+		unregister_post_type( 'with_interface_kind' );
+		unregister_post_type( 'with_union_kind_one' );
+		unregister_post_type( 'with_union_kind_two' );
+	}
+
+	public function testRegisterCustomPostTypeWithInterfaceKind() {
+		register_post_type( 'with_interface_kind', [
+			'public'               => true,
+			'show_in_graphql'      => true,
+			'graphql_single_name'  => 'WithInterfaceKind',
+			'graphql_plural_name'  => 'WithInterfaceKinds',
+			'graphql_kind'         => 'interface',
+			'graphql_resolve_type' => $this->resolve_type(),
+		]);
+
+		register_post_type( 'child_type_one', [
+			'public'              => true,
+			'show_in_graphql'     => true,
+			'graphql_single_name' => 'ChildTypeOne',
+			'graphql_plural_name' => 'ChildTypeOne',
+			'graphql_interfaces'  => [ 'WithInterfaceKind' ],
+		]);
+
+		register_post_type( 'child_type_two', [
+			'public'              => true,
+			'show_in_graphql'     => true,
+			'graphql_single_name' => 'ChildTypeTwo',
+			'graphql_plural_name' => 'ChildTypeTwo',
+			'graphql_interfaces'  => [ 'WithInterfaceKind' ],
+		]);
```

Comment:
This is actually a pretty confusing experience. 

Here we've registered 3 post types, and will get 3 post types in the admin. 

But, we've set the `graphql_kind` for `WithInterfaceKind` to be an `interface`, and the `child_type_one` and `child_type_two` to implement that interface. 

When we create posts of the `WithInterfaceKind` in the wp-admin, we have no way to query for them, because the `{ withInterfaceKinds { nodes { ... } } }` query tries to resolve posts of the `ChildTypeOne` and `ChildTypeTwo` types 🤔 


How are we supposed to use data from the `WithInterfaceKind` post type? 🤔

Reply:
Actually, it's trying to resolve posts from the database of the `with_interface_kind` post_type, but then trying to resolve the type as `ChildTypeOne` or `ChildTypeTwo`, but these types won't ever be returned from the database because the resolver is executing this query: 

```
SELECT wp_posts.ID
FROM wp_posts
WHERE 1=1  AND wp_posts.post_type = 'with_interface_kind' AND ((wp_posts.post_status = 'publish')) 
ORDER BY wp_posts.post_date DESC, wp_posts.ID DESC
LIMIT 0, 11
```

So, we're always getting a list of `with_interface_kind` posts, so the `resolveType` function will try and return the `WithInterfaceKind` Type, but the allowed return types are `ChildTypeOne` and `ChildTypeTwo` which will never be returned by the underlying query 🤔 


So we'll always end up with errors like the following:

![CleanShot 2022-09-09 at 14 14 18](https://user-images.githubusercontent.com/1260765/189435688-7693ee02-7903-4b49-a957-609d4d0bab96.png)

Reply:
The original plan was to handle this case (vs a more innocuous one where the the same post type can have different GraphQL types - e.g a Post Format, or TEC Event, or WC_Product type) with #2466 . 
Now that it's a separate issue, I need to rewrite this test to check the above simpler use case.

Reply:
I think in the case of WooCommerce, the "variations" etc are still of the same "product" post_type, but in this example we're trying to use 3 different post types and the resolver logic doesn't make that connection. 

I think for WooCommerce, we're trying to get to a point where you can do something like: 

```graphql
{
  products { # query from the "product" post type
     nodes {
        __typename
        id
        ...on VariableProduct { someField } # "differentiate type based on some value of the product"
        ...on ProductBundle { someOtherField } # "differentiate type based on some value of the product"
     }
  }
}
```

Where in the test example, we're actually registering 3 different post types, but the resolver is still trying to resolve `WithInterfaceKind` from the `with_interface_kind` post_type. 

I think we might want to test with a scenario like so: 

```php
add_action( 'init', function() {

	register_post_type( 'with_interface_kind', [
		'public'               => true,
		'show_in_graphql'      => true,
		'supports' => [ 'title', 'editor', 'custom-fields', 'author' ],
		'label'                => 'WithInterfaceKind',
		'graphql_single_name'  => 'WithInterfaceKind',
		'graphql_plural_name'  => 'WithInterfaceKinds',
		'graphql_kind'         => 'interface',
		'graphql_resolve_type' => function ( $post ) {

			if ( 'true' === get_post_meta( $post->databaseId, 'is_child_one', true ) ) {
				$kind = 'ChildTypeOne';
			} else {
				$kind = 'ChildTypeTwo';
			}

			return $kind;
		},
	]);

	register_graphql_object_type( 'ChildTypeOne', [
		'eagerlyLoadType' => true,
		'interfaces' => [ 'WithInterfaceKind' ],
		'fields' => [
			'childTypeOneField' => [
				'type' => 'String',
				'resolve' => function( $post ) {
					return get_post_meta( $post->databaseId, 'child_one_field', true );
				}
			]
		]
	]);

	register_graphql_object_type( 'ChildTypeTwo', [
		'eagerlyLoadType' => true,
		'interfaces' => [ 'WithInterfaceKind' ],
		'fields' => [
			'ChildTypeTwoField' => [
				'type' => 'String',
				'resolve' => function( $post ) {
					return get_post_meta( $post->databaseId, 'child_two_field', true );
				}
			]
		]
	]);

} );
```

Here, we register the Post Type as `'graphql_kind' => "interface"`, but instead of registering new post types that implement that interface, we register new Object Types that will implement that interface. 

The idea with the Post Type being an interface is that there can be variables within posts of that post_type that can make them different types, but ultimately they're all still "WithInterfaceKind" types.

The `resolveType` function is now not determining Type by the node's `post_type`, but another differentiating factor. 

We should be confident that registering a post_type will always return a node of that post_type, but the final Type could be something else based off a meta value, a status, a taxonomy relationship, etc. . .but it will always be of the same Post Type (the Interface being the common bond that the object types implement)

Reply:
That would allow for this behavior: 

![CleanShot 2022-09-09 at 14 39 28](https://user-images.githubusercontent.com/1260765/189443529-c512bc7e-e01b-439c-85de-d7990eaf236a.png)

Reply:
It might be good to figure out a way to guard against one post_type implementing an `Interface` that represents _another_ post_type. 

Unless there's a strong case for it, I don't see this as something we want to support. 

Post Types are, in my mind, different objects altogether. 

`PostTypeA` should never implement an interface for `PostTypeB`. 

I'm open for examples where that does sound like a good idea though. I could be thinking too narrowly. 

What might be good is to output a `graphql_debug()` message if that happens, letting the user know that the post type attempted to implement an interface represented by another post type, and that's not the intended use. 🤷🏻‍♂️

Reply:
> It might be good to figure out a way to guard against one post_type implementing an Interface that represents another post_type.
>
>Unless there's a strong case for it, I don't see this as something we want to support.

This is a good point. I think I was focused more on how we _could_  (via #2466 ) instead of whether we should.

Even if we do think of a valid use case, since itl require #2466 to implement, we can always revisit the decision then and worse case roll-back the protection.

Reply:
Ya, we'd need to go even deeper than #2466 I think. We'd need a way for the entire Connection Resolver to be overridden. 

Like 

```php
register_post_type( 'interface_type', [
  'show_in_graphql' => true,
  'graphql_single_name' => 'IntefaceType',
  'graphql_plural_name' => 'InterfaceTypes',
  'graphql_connection_args' => function( $args ) { 
     # we'd need a way to override the connection args to query across different post_types
     # this post_type should implement the Interface defined by a different post_type
     $args['post_type'] = [ 'interface_type', 'post_type_a', 'post_type_b' ];
     return $args;
  }
  # we'd also need to customize the loader
  ...
  # and the model
  ...
  # and probably some other things
] );

register_post_type( 'post_type_a', [
  'graphql_interfaces' => [ 'InterfaceType' ], # this post_type should implement the Interface defined by a different post_type
  'show_in_graphql' => true,
  'graphql_single_name' => 'PostTypeA',
  'graphql_plural_name' => 'PostTypeAs',

] );

register_post_type( 'post_type_b', [
  'graphql_interfaces' => [ 'InterfaceType' ], # this post_type should implement the Interface defined by a different post_type
  'show_in_graphql' => true,
  'graphql_single_name' => 'PostTypeB',
  'graphql_plural_name' => 'PostTypeBs',
] );
```

Reply:
tests fixed in https://github.com/wp-graphql/wp-graphql/pull/2399/commits/9a0c7668dba53b10b2f1a8730ee891b98eb9e670m